lib/Sema/CSStep.cpp

```diff
@@ -525,29 +528,132 @@ StepResult DisjunctionStep::resume(bool prevFailed) {
+                           IsDeclSubstitutableRequest{ declA, declB },
+                           false);
+}
+
 bool DisjunctionStep::shouldSkip(const DisjunctionChoice &choice) const {
   auto &ctx = CS.getASTContext();
 
-  bool attemptFixes = CS.shouldAttemptFixes();
-  // Enable all disabled choices in "diagnostic" mode.
-  if (!attemptFixes && choice.isDisabled()) {
+  // Never skip disjunction choices in diagnostic mode.
+  if (CS.shouldAttemptFixes())
+    return false;
+
+  auto skip = [&](std::string reason) -> bool {
     if (CS.isDebugMode()) {
       auto &log = getDebugLogger();
-      log << "(skipping ";
+      log << "(skipping " + reason + " ";
       choice.print(log, &ctx.SourceMgr);
       log << '\n';
     }
 
     return true;
-  }
+  };
 
-  // Skip unavailable overloads unless solver is in the "diagnostic" mode.
-  if (!attemptFixes && choice.isUnavailable())
-    return true;
+  if (choice.isDisabled())
+    return skip("disabled");
+
+  // Skip unavailable overloads.
+  if (choice.isUnavailable())
+    return skip("unavailable");
 
   if (ctx.TypeCheckerOpts.DisableConstraintSolverPerformanceHacks)
     return false;
 
+  // If the solver already found a solution with a better overload choice that
+  // can be unconditionally substituted by the current choice, skip the current
+  // choice.
+  if (LastSolvedChoice && LastSolvedChoice->second == getCurrentScore() &&
+      choice.isGenericOperator()) {
+    auto *declA = LastSolvedChoice->first->getOverloadChoice().getDecl();
+    auto *declB = static_cast<Constraint *>(choice)->getOverloadChoice().getDecl();
+
+    if (TypeChecker::compareDeclarations(CS.DC, declA, declB) == Comparison::Better) {
+      if (isDeclSubstitutable(declA, /*by=*/declB))
+        return skip("subtype");
+    }
+  }
+
+  // If the solver already found a solution with a choice that did not
+  // introduce any conversions (i.e., the score is not worse than the
+  // current score), we can skip any generic overloads with conformance
+  // requirements that are not satisfied by any known argument types.
+  auto bestScore = getBestScore(Solutions);
+  auto bestChoiceNeedsConversions = bestScore && (bestScore > getCurrentScore());
+  if (bestScore && !bestChoiceNeedsConversions && choice.isGenericOperator() && argFnType) {
+    Constraint *constraint = choice;
+    auto *decl = constraint->getOverloadChoice().getDecl();
+    auto *useDC = constraint->getOverloadUseDC();
+    auto choiceType = CS.getEffectiveOverloadType(constraint->getOverloadChoice(),
+                                                  /*allowMembers=*/true, useDC);
+    auto choiceFnType = choiceType->getAs<FunctionType>();
+    auto genericFnType = decl->getInterfaceType()->getAs<GenericFunctionType>();
+    auto signature = genericFnType->getGenericSignature();
+
+    for (auto argParamPair : llvm::zip(argFnType->getParams(),
+                                       choiceFnType->getParams())) {
```

Comment:
This is only done for operator calls, which to my knowledge can't have default parameters...right?

Reply:
Sorry yeah, comment speaks about generic overloads so I didn't notice that it was isGeneric*Operator*() in the check...

Reply:
Whoops, I will fix up the comment, thanks! I had originally written some of this thinking it would work for all generic overloads, and then realized why it wouldn't haha